### PR TITLE
Fix early-data wakeup loss

### PR DESCRIPTION
**Pull request by quininer**
Fix https://github.com/tokio-rs/tls/issues/54

### CODE REVIEW

File: tokio-rustls/src/client.rs [hunk 1]
```diff
@@ -58,7 +61,24 @@ where
     ) -> Poll<io::Result<()>> {
         match self.state {
             #[cfg(feature = "early-data")]
-            TlsState::EarlyData(..) => Poll::Pending,
+            TlsState::EarlyData(..) => {
+                let this = self.get_mut();
+
+                // At this time, we have not really established a tls connection,
```
Comment (djc) [inline tokio-rustls/src/client.rs:67]: Thanks for adding this, that's helpful! I guess the other bit of information that might be useful to mention here is that in the `EarlyData` state, we are waiting for the `AsyncWrite` impl to transition us to `Stream`, so we ignore readiness events from the read side until the write side says we're done writing the early data. Maybe make that more explicit in the comment?

File: tokio-rustls/src/client.rs [hunk 2]
```diff
@@ -136,6 +156,11 @@ where
 
                 // end
                 this.state = TlsState::Stream;
+
+                if let Some(waker) = this.early_waker.take() {
+                    waker.wake();
+                }
```
Comment (djc) [inline tokio-rustls/src/client.rs:162]: It looks like there's quite a bit of code duplication between the early data handling in `poll_write()` and `poll_flush()` maybe pull that out into a separate method? (Ideally in a separate commit.)

Comment (quininer) [reply tokio-rustls/src/client.rs:161]: agree, we can complete it in a later PR.

File: tokio-rustls/src/client.rs (hunk 1 above)
Comment (quininer) [reply tokio-rustls/src/client.rs]: I updated the comment.

**Review by djc — APPROVED** [review]

**Pull request merged**